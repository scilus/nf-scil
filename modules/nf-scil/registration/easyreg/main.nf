

process REGISTRATION_EASYREG {
    tag "$meta.id"
    label 'process_single'

    container "freesurfer/freesurfer:7.4.1"

    input:
    tuple val(meta), path(reference), path(floating), path(ref_segmentation), path(flo_segmentation)

    output:
    tuple val(meta), path("*_reference_segmentation.nii.gz") , emit: ref_seg
    tuple val(meta), path("*_floating_segmentation.nii.gz")  , emit: flo_seg
    tuple val(meta), path("*_reference_registered.nii.gz")   , emit: ref_reg
    tuple val(meta), path("*_floating_registered.nii.gz")    , emit: flo_reg
    tuple val(meta), path("*_forward_field.nii.gz")          , emit: fwd_field, optional: true
    tuple val(meta), path("*_backward_field.nii.gz")         , emit: bak_field, optional: true
    path "versions.yml"                                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def field = task.ext.field ? "--fwd_field ${prefix}_forward_field.nii.gz --bak_field ${prefix}_backward_field.nii.gz " : ""
    def threads = task.ext.threads ? "--threads " + task.ext.threads : ""
    def affine = task.ext.affine ? "--affine_only " : ""

    """
    if [[ -f "$ref_segmentation" ]];
    then
        reference_segmentation=$ref_segmentation
    else
        reference_segmentation="${prefix}_reference_segmentation.nii.gz"
    fi

        if [[ -f "$flo_segmentation" ]];
    then
        floating_segmentation=$flo_segmentation
    else
        floating_segmentation="${prefix}_floating_segmentation.nii.gz"
    fi

    mri_easyreg --ref $reference --flo $floating --ref_seg \${reference_segmentation} --flo_seg \${floating_segmentation} \
        --flo_reg ${prefix}_floating_registered.nii.gz --ref_reg ${prefix}_reference_registered.nii.gz $field $threads $affine

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4.1
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mri_easyreg -h

    touch ${prefix}_reference_segmentation.nii.gz
    touch ${prefix}_floating_segmentation.nii.gz
    touch ${prefix}_reference_registered.nii.gz
    touch ${prefix}_floating_registered.nii.gz
    touch ${prefix}_forward_field.nii.gz
    touch ${prefix}_backward_field.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4.1
    END_VERSIONS
    """
}
