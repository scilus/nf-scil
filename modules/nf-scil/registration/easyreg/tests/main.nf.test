nextflow_process {

    name "Test Process REGISTRATION_EASYREG"
    script "../main.nf"
    process "REGISTRATION_EASYREG"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/easyreg"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "others.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - easyreg") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                        .map{ test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_crop.nii.gz"),
                        file("\${test_data_directory}/t1_resample.nii.gz"),
                        [],
                        []
                ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
