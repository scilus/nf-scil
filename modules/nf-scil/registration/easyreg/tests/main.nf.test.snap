{
    "registration - easyreg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_reference_segmentation.nii.gz:md5,aeac0f1cf19381a7f40464408dc6b9e9"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_floating_segmentation.nii.gz:md5,c895e33807532642f197806b230aaa79"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_reference_registered.nii.gz:md5,cfaa3b3c6d30d25ef0b8dd5c6ecce0e4"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_floating_registered.nii.gz:md5,817e5d3913acdfe1176889ee8566f714"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_forward_field.nii.gz:md5,76e3ffa48bf6ba1b97d7e782e20686b8"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_backward_field.nii.gz:md5,99b2624aecaa1d0fe3e0633eab433344"
                    ]
                ],
                "6": [
                    "versions.yml:md5,4661210880c42a986923e8257f64c760"
                ],
                "bak_field": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_backward_field.nii.gz:md5,99b2624aecaa1d0fe3e0633eab433344"
                    ]
                ],
                "flo_reg": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_floating_registered.nii.gz:md5,817e5d3913acdfe1176889ee8566f714"
                    ]
                ],
                "flo_seg": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_floating_segmentation.nii.gz:md5,c895e33807532642f197806b230aaa79"
                    ]
                ],
                "fwd_field": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_forward_field.nii.gz:md5,76e3ffa48bf6ba1b97d7e782e20686b8"
                    ]
                ],
                "ref_reg": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_reference_registered.nii.gz:md5,cfaa3b3c6d30d25ef0b8dd5c6ecce0e4"
                    ]
                ],
                "ref_seg": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_reference_segmentation.nii.gz:md5,aeac0f1cf19381a7f40464408dc6b9e9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4661210880c42a986923e8257f64c760"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-25T20:31:22.056733"
    }
}